File: programmers.playground/Sources/베스트 앨범.swift
// 베스트 앨범
// 2023.03.14
// https://school.programmers.co.kr/learn/courses/30/lessons/42579

import Foundation

func solution(_ genres:[String], _ plays:[Int]) -> [Int] {
    var bestAlbum : [String: [[Int]]] = [:]
    var result : [Int] = []
    
    for i in 0..<plays.count{
        if bestAlbum[genres[i]] != nil{
            bestAlbum[genres[i]]?.append([i, plays[i]])
        } else {
            bestAlbum[genres[i]] = [[i,plays[i]]]
        }
    }
    
    let albumSorted = bestAlbum.sorted {
        return $0.value.map { $0[1] }.reduce(0, +) > $1.value.map { $0[1] }.reduce(0, +)
    }
    
    for i in 0..<albumSorted.count{
        let a = albumSorted[i].value.sorted{$0[1] > $1[1]}
        
        for j in 0..<a.count{
            if j > 1{ break }
            result.append(a[j][0])
        }
    }
    
    return result
}

File: Programmers/베스트앨범.py
# 베스트앨범
# 2024.12.22
# https://school.programmers.co.kr/learn/courses/30/lessons/42579
# Not Record

def solution(genres, plays):
    answer = []
    dic = {}
    for i in range(0, len(genres)):
        if genres[i] in dic:
            dic[genres[i]].append([plays[i], i])
        else:
            dic[genres[i]] = [[plays[i], i]]

    sorted_dic = sorted(dic.values(), key= lambda x: sum(i[0] for i in x), reverse = True)
    
    for value in sorted_dic:
        answer += [i[1] for i in sorted(value, key = lambda x:x[0], reverse=True)[:2]]

    return answer


'''
코드를 더 함축하지 못했다.
다른 사람들 풀이를 보니까 푸는 방식은 비슷한대 함축해서 코드를 작성했다.

굳이 조건문 없이 dic을 선언할떄도
dic = {e:[] for e in set(genres)}
해서 음악 종류별로 있는 dictionary를 만들고

for e in zip(genres, plays, range(len(plays))):
    dic[e[0]].append([e[1], e[2]])

를 이용해서 나랑 다르게 반복문을 줄였다.

조금 더 코테문제 풀다보면 나도 저런 응용이 가능할꺼같다.
'''